src/com/amazon/ion/impl/_Private_RecyclingStack.java

```diff
@@ -0,0 +1,95 @@
+package com.amazon.ion.impl;
+
+import java.util.ArrayList;
+import java.util.List;
+
+/**
+ * A stack whose elements are recycled. This can be useful when the stack needs to grow and shrink
+ * frequently and has a predictable maximum depth.
+ * @param <T> the type of elements stored.
+ */
+public final class _Private_RecyclingStack<T> {
```

Comment:
It seems like another factoring of this would be pairing a general-purpose object pool with a stack. I guess @zslayton already pointed this out in [this conversation from #369](https://github.com/amzn/ion-java/pull/369#discussion_r657368163).

Nothing to do here, for now.

Reply:
+1